Euclid beat24 futurebeat

modulove opened the pull request:
This is the feature Update 2024 "futurebeat" - ARYTHMATIK_Euclid.ino

Using FastGPIO now
Added savestate function (WIP)
Fixes  and additions for Menue and UI  (WIP)
New quick access Menue with Toggle Mute (rotate encoder CCW) (WIP)
Its possible now to reset single channels and introduce phasing polyrythmic patterns (polymeter?)
New Sequence reset function 
Hits, Offset & Length can be dialed in CW & CCW

earsneyes committed:
Update ARYTHMATIK_Euclid.ino

Added all draft functions, needs cleanup still

earsneyes committed:
Update ARYTHMATIK_Euclid.ino

flip display again

earsneyes committed:
Fix savestate ARYTHMATIK_Euclid.ino

Cleaned up some mess.
Lots of in there still

earsneyes committed:
Update ARYTHMATIK_Euclid.ino

Fixes for menue & some more cleanup.

earsneyes committed:
Update ARYTHMATIK_Euclid.ino

Fixed Menue
New Quick access Menue (rotate encoder CCW) with Toggle Mute

modulove committed:
Merge pull request #5 from earsneyes/patch-1

Feature Update 2024 "futurebeat" - ARYTHMATIK_Euclid.ino

Using FastGPIO now
Added savestate function (WIP)
Fixes for Menue and UI (WIP)
New quick access Menue with Toggle Mute (rotate encoder CCW) (WIP)
Its possible now to reset single channels and introduces phasing patterns (polymeter?)
New Sequence reset function 
Hits, Offset & Length can be dialed in CW & CCW finally

modulove committed:
Update ARYTHMATIK_Euclid.ino

Small UI improvements

modulove committed:
Update ARYTHMATIK_Euclid.ino

Add UPSIDEDOWN define for the new flip-panel

modulove committed:
Update ARYTHMATIK_Euclid.ino

Whooops. CLK in & RST need to be switched as well.

awonak commented on Firmware/ARYTHMATIK_Euclid/ARYTHMATIK_Euclid.ino:
```diff
@@ -29,296 +66,411 @@ long interval = 1000;
-Encoder myEnc(3, 2);//use 3pin 2pin
-int oldPosition  = -999;
+// define for using module upside down in your case
+#define UPSIDEDOWN
```
Did you intend to leave the `UPSIDEDOWN` flag enabled? I would imagine you'd want the default repository version to not have this enabled by default.

awonak commented on Firmware/ARYTHMATIK_Euclid/ARYTHMATIK_Euclid.ino:
```diff
@@ -29,296 +66,411 @@ long interval = 1000;
   }
-  if (select_ch == 6 && select_menu < 0) { // random mode
-    select_menu = 1;
+  if (select_ch == 6 && select_menu < 0) {  // random mode  //Begrenzer
```
[optional] To help with readability, consider using an Enum for the selected menu page and any enumerated list of distinct choices. For example:
```c
// Enum constants for current display page.
enum MenuPage {
    PAGE_HITS,
    PAGE_OFFSET,
    PAGE_LENGTH,
    PAGE_MUTE,
    ...
};
MenuPage selected_menu = PAGE_HITS;
```
Then reading the code will be much cleaner:
```c
if (select_ch >= 7 && selected_page == PAGE_HITS) {
    ...
}
if (select_ch >= 7 && selected_page == PAGE_OFFSET) {
    ...
}
...
```
Currently, it's difficult for the reader to understand the meaning of statements like "`select_ch == 6 && select_menu < 0`" versus "`selected_ch == 6 && selected_menu != PAGE_RANDOM`"

awonak commented on Firmware/ARYTHMATIK_Euclid/ARYTHMATIK_Euclid.ino:
```diff
@@ -29,296 +66,411 @@ long interval = 1000;
+      case 1:                  //hits
+        if (select_ch != 6) {  // not random mode
+          hits[select_ch]--;
+          if (hits[select_ch] >= 17) {  //Begrenzer nach unten durch Rücklauf "unter Null", d. h. auf 255 wegen vorzeichenloser Byte-Definition
```
[optional] I'd suggest standardizing comments to match the code language in English.

modulove replied:
This was leftover and is now done.

awonak commented on Firmware/ARYTHMATIK_Euclid/ARYTHMATIK_Euclid.ino:
```diff
@@ -29,296 +66,411 @@ long interval = 1000;
-        limit[select_ch]-- ;
-        if (limit[select_ch] >= 17) {
-          limit[select_ch] = 0;
+      case 3:  // there seemed to be a bug with the playing indicatior dot hiding but cant spot it now
```
Interesting, I see that a bug was removed, because it was previously decrementing `limit[selected_ch]` and then erroneously checking if the updated value was _greater_ than or equal to 17 :)

awonak commented on Firmware/ARYTHMATIK_Euclid/ARYTHMATIK_Euclid.ino:
```diff
@@ -29,296 +66,411 @@ long interval = 1000;
+      case 3:  // there seemed to be a bug with the playing indicatior dot hiding but cant spot it now
+        limit[select_ch]--;
+        if (limit[select_ch] <= 0) {
+          limit[select_ch] = 17;
```
I believe this code is enforcing a lower bounds limit when decrementing this value, so when it is decremented below zero, we want to set it back to zero, not 17. Furthermore, if `limit[select_ch]` is an unsigned int, decrementing below zero will wrap the value back around to the highest value. For example, a uint_8 of 0 - 1 will result in 255. You should check if the value is zero before decrementing.

Suggested fix:
```c
if (limit[select_ch] > 0) {
   limit[select_ch]--;
}
```

awonak commented on Firmware/ARYTHMATIK_Euclid/ARYTHMATIK_Euclid.ino:
```diff
@@ -327,69 +479,119 @@ void loop() {
+        playing_step[i] = 0;  // step limit is reached
+      }
+    }
+    #ifdef UPSIDEDOWN
```
There is an opportunity to simplify these changes by using the `#ifdef UPSIDEDOWN` flag to assign each output to a variable and use that instead of checking the `UPSIDEDOWN` flag each time. For example, add this definition at the top of your script:
```c
#ifdef UPSIDEDOWN
#define OUTPUT1 FastGPIO::Pin<8>
#define LED1 FastGPIO::Pin<0>
...
#else
#define OUTPUT1 FastGPIO::Pin<5>
#define LED1 FastGPIO::Pin<14>
...
#endif
```
Then use them like this without needing to check `#ifdef UPSIDEDOWN`:
```c
switch (k) {
  case 0:  //CH1
    OUTPUT1::setOutput(1);
    LED1::setOutput(1);
    break;
  ...
}
```

awonak commented on Firmware/ARYTHMATIK_Euclid/ARYTHMATIK_Euclid.ino:
```diff
@@ -398,133 +600,642 @@ void loop() {
+  int baseAddress = EEPROM_START_ADDRESS + (slot * CONFIG_SIZE);
+
+  for (int ch = 0; ch < 6; ++ch) {
+    EEPROM.put(baseAddress + ch * sizeof(byte), memorySlots[slot].hits[ch]);        // hits
```
Why not store the whole struct? Then on load pick the request channel from the full loaded value.
```c
EEPROM.put(0, memorySlots);
```

modulove replied:
done.

awonak reviewed (commented):
Excellent improvements so far! I've added a few notes from my first pass. I didn't get through reviewing the entire script, but I have added some initial notes that should be helpful in cleaning up the code to make it more readable and reduce the number of lines.

Additional notes:
- There are quite a lot of variable definitions that this app uses. I would suggest spending some time to logically group them together to make it more clear what variables are present and used in what context. Also group your `#DEFINE` at the top, followed by const definitions, then variables after that. Any value a developer might edit should be up top.
- I think it's a good idea to add SimpleRotary here, there are a lot of variables and logic related to reading the encoder, which distracts from the actual firmware app logic. (out of scope for this PR, but consider using my hardware abstraction library for A-RYTH-MATIK - https://github.com/awonak/libModulove, though it doesn't support UPSIDEDOWN display yet...)
- I love that you've started using the SlotConfiguration struct for organizing each channel's variables. I think the script could be improved further by using the struct for both persistent storage and for script logic. Then you keep / access all of your references in one place.

modulove committed:
Update ARYTHMATIK_Euclid.ino

prepare for ENCODER_REVERSED

modulove committed:
Update ARYTHMATIK_Euclid.ino

Add PANEL_USD use

modulove committed:
Update ARYTHMATIK_Gate-seq.ino

ENCODER_REVERSED & PANEL_USD

modulove committed:
Update ARYTHMATIK_Euclid.ino

Default use is not PANEL_USD

modulove committed:
Update ARYTHMATIK_Euclid.ino

remove comment

modulove committed:
Update ARYTHMATIK_Euclid.ino

ifdef ENCODER_REVERSED instead of #ifdef INVERTED

modulove committed:
Rename hardware_test.ino to Firmware/DEBUG/DEBUG.ino

Move hw test sketch to Firmware/DEBUG/DEBUG.ino for github actions

modulove committed:
Merge branch 'main' into euclid-beat24-futurebeat

modulove committed:
Delete Firmware/DEBUG directory

modulove committed:
Update ARYTHMATIK_Euclid.ino

Housekeeping comments, config, few small fixes.
This should make the code a little more readable in some places.
Will have another take on the remaining changes soon.

modulove committed:
Update ARYTHMATIK_Euclid.ino

Running into some issues here so keep commits small.
Changed flip Panel defines.
Added loading default settings (WIP)

modulove committed:
Update ARYTHMATIK_Euclid.ino

Simplify ROTATE_PANEL

modulove committed:
Update ARYTHMATIK_Euclid.ino

Make this work actually. 
Remember remember.. test with actually fully working hardware!

modulove committed:
Update ARYTHMATIK_Euclid.ino

Use SimpleRotary

modulove committed:
Update ARYTHMATIK_Euclid.ino

Reintroduce bar_select manipulation in random mode allowing adjustment with rotary encoder.

modulove committed:
Update ARYTHMATIK_Euclid.ino

Some smaller improvements for the UI and some other places.
This one uses still SimpleRotary, working on replacing it and cleaning up the implementation.

modulove committed:
Update README.md

Add iBom preview link

modulove committed:
Update ARYTHMATIK_Euclid.ino

Replaced SimpleRotary with EncoderButton
Added probability & UI
More small UI changes

modulove committed:
Update ARYTHMATIK_Euclid.ino

Progress in onEncoderClicked. Still no saveConfig, loadConfig. Need to tackle the menu structure

modulove committed:
Update ARYTHMATIK_Euclid.ino

test encoder acceleration
some more smaller fixes

modulove committed:
Update ARYTHMATIK_Euclid.ino

No TapTempo yet

modulove committed:
Update ARYTHMATIK_Euclid.ino

Added onEncoderPressedRotation (Long button press) for the ability to quickly dial in Hits from the first level of the menue (Select channel) while the button is pressed.

modulove committed:
Update ARYTHMATIK_Euclid.ino

Add "MUTE" display when all channels are mutet.

modulove committed:
Update ARYTHMATIK_Euclid.ino

Fix Random_change() so it advances the channels with encoder rotation.

modulove committed:
Update ARYTHMATIK_Euclid.ino

Fix initializeCurrentConfig

modulove committed:
Update ARYTHMATIK_Euclid.ino

Last changes before superbooth. Yay

awonak committed:
starting work on enums

awonak committed:
initial enum work

modulove committed:
Update ARYTHMATIK_Euclid.ino

squash menu changes

modulove committed:
Update ARYTHMATIK_Euclid.ino

Rework channel & select menu to make it more readable. looks pretty ok to me.  in depth testing needed still

bkrsmdesign committed:
Update ARYTHMATIK_Euclid.ino

added the boot logo in the config and setup sections

bkrsmdesign committed:
Update ARYTHMATIK_Euclid.ino

added comments to changes

bkrsmdesign committed:
Update ARYTHMATIK_Euclid.ino

reversed the logic of offset so it rotates in the right direction

bkrsmdesign committed:
Update ARYTHMATIK_Euclid.ino

RSET

modulove committed:
Merge pull request #13 from bkrsmdesign/euclid-beat24-futurebeat

Euclid beat24 futurebeat boot logo

modulove committed:
ARYTHMATIK_Euclid.ino - Refactor onEncoderClicked for readability

Changes from @awonak

modulove committed:
Turn off clock LED

Turn off clock LED along with the output pins after a delay of 10 milliseconds.

modulove committed:
Some cleanup - ARYTHMATIK_Euclid.ino

cleanup

modulove committed:
Adjust Hits for channel 6 as well

modulove committed:
RandomAdvance & Random X mode

Make these work again while holding button & rotating Encoder

modulove committed:
ARYTHMATIK_Euclid.ino - More cleanup and fixes

Select EEPROM Slot with held encoder (WIP)
Several performance improvements

modulove committed:
ARYTHMATIK_Euclid.ino - (onEncoderPressedRotation) select channel in Channel edit modes

encoder's pressed rotation can now select the channel when editing Channel settings (Hits,Offset,Limit,Mute,Random,Prob). Making the UI more intuitive and functional.

modulove committed:
ARYTHMATIK_Euclid.ino - some quick patches 

Updated default (preset) config for presets with names (WIP)
Replaced TEMPO with PRESET menu (Internal Clock later)

awonak committed:
mostly minor updates:

- create const for common magic numbers
- add CLK_LED def
- remove unnecessary rotate 0 checks
- use increment, not accelerated for random advance

modulove committed:
Merge pull request #17 from awonak/awonak-euclid-beat24

[WIP] awonak-euclid-beat24

awonak committed:
Merge branch 'heads/upstream/euclid-beat24-futurebeat' into awonak-euclid-beat24

awonak committed:
Fixed behavior for showing menu overlay during encoder press and hold for advanced global settings like Save, Load, and Preset.

awonak committed:
minor global var refactor.

modulove committed:
Merge pull request #18 from awonak/awonak-euclid-beat24

Push & Hold menu refactor

awonak committed:
This will update how the UI Refresh behavior works:

When clock running, refresh on:
- clock input
- reset input
- encoder press / release
- NOT on encoder rotate (one of the above must happen to update UI)

When not running, refresh on:
- any of the events above
- after 1 second (a half note at 120bpm), UI constantly updates

awonak committed:
Fixed index bug causing all channesl to follow ch 1.

modulove committed:
Merge pull request #19 from awonak/awonak-euclid-beat24

A few more minor fixes

modulove committed:
ARYTHMATIK_Euclid.ino - load the latest saved state on boot

Load the latest saved state on boot and save the last slot used to EEPROM

awonak committed:
Allow encoder to refresh the display with a throttled min refresh rate.

awonak committed:
improved refresh rate check

awonak committed:
fixed bug in logic. allows even faster ui refresh rate.

modulove committed:
Merge pull request #20 from awonak/awonak-euclid-beat24

Improve the UI refresh rate

modulove committed:
Update ARYTHMATIK_Euclid.ino

modulove committed:
ARYTHMATIK_Euclid.ino - remove artifacts

Only one text size for now
Cleanup rythms (WIP)

modulove committed:
ARYTHMATIK_Euclid.ino - Update gate output

modulove committed:
ARYTHMATIK_Euclid.ino remove CLOCK_STOP_DURATION 

remove CLOCK_STOP_DURATION

awonak committed:
Mostly code cleanup, no decent performance improvements yet.
* Consistent side menu drawing functions.
* Fully separate the unique Probability screen from the Euclid Pattern screen
* Replace F() string with char when printing single character
* Consistant use of const for MAX_CHANNELS and MAX_STEPS

modulove committed:
Merge pull request #21 from awonak/awonak-euclid-beat24

Mostly code cleanup, no decent performance improvements yet.

awonak committed:
Fixed logic bug in display refresh rate.

awonak committed:
Merge branch 'euclid-beat24-futurebeat' of https://github.com/modulove/A-RYTH-MATIK into awonak-euclid-beat24

modulove committed:
encoder.setRateLimit(20);

This was needed for me to make it not skip ouptuts while using the encoder. Now works as discribed without missing here as well

awonak committed:
increase min refresh duration

100ms was still causing beats to skip, so I increased it up to 250ms and I can no longer detect any missed beats or wonky behavior.

modulove committed:
ARYTHMATIK_Euclid.ino - Fix for toggle Mute & Don't mute when manually advancing in "X" Random mode

Fix for toggle Mute function
Don't mute when manually advancing in "X" Random mode

modulove committed:
ARYTHMATIK_Euclid.ino - Update patterns

Test pattern, more consistent (valid) rhythms (WIP)

modulove committed:
ARYTHMATIK_Euclid.ino - restore onEncoderPressedRotation & Fix external clock detection

restore onEncoderPressedRotation for quick edit hits,offset,limit, .. while in edit modes
- Fix clock detection so it switches between internal & clock input

modulove committed:
ARYTHMATIK_Euclid.ino - Update onEncoderRotation

Fix Bug with remove adjusting Hits in onEncoderRotation

modulove committed:
ARYTHMATIK_Euclid.ino - Update onEncoderPressedRotation

Use Random_change() and Random_change_mute() functions  based on the encoder rotation direction while encoder button is pressed

modulove committed:
ARYTHMATIK_Euclid.ino - fix Display bug when only 1 hit is defined

The hits line was mirrored when there is only 1 hit (Thanks Sasha)

modulove committed:
ARYTHMATIK_Euclid.ino - Clock range & bootlogo

Internal clock range 20 to 280 BPM
Disable bootlogo animation if the CPU frequency is 32 MHz

modulove committed:
ARYTHMATIK_Euclid.ino - small UI changes

Draw hits info if > 9 hits
Always draw "Mute M" per channel
Remove Limit "L"

modulove committed:
ARYTHMATIK_Euclid.ino - Display BPM if external clock

Display BPM if external clock

modulove committed:
ARYTHMATIK_Euclid.ino - Longpress to enable/disable internal clock

awonak committed:
Calculate period along with tempo change.

awonak committed:
Fixed a bug clock bug

When internal clock mode active and an external clock was sent to CLK, the period would get double counted. Now we check clock input according to internalClock value.

Also removed some dead code that was not executing (ch select happens in handleSettingNavigation).

modulove committed:
ARYTHMATIK_Euclid.ino - Save/Load BPM & Clock state

Save & Load Clock state and Tempo
Load Tempo from presets

modulove committed:
ARYTHMATIK_Euclid.ino - Update Clock UI  & Mode

Update Clock UI  & Mode
Add divider for external clock mode

modulove committed:
ARYTHMATIK_Euclid.ino - Remove clock divider feature

Remove clock divider feature

modulove committed:
Update ARYTHMATIK_Euclid.ino - few smaller fixes

Refresh hits at boot time
Combine Random change functions

modulove committed:
Update ARYTHMATIK_Euclid.ino - Add overlay menu timeout

Add overlay menu timeout

modulove committed:
Update ARYTHMATIK_Euclid.ino - Use encoderButton timeout

Use encoderButton timeout

modulove committed:
Update ARYTHMATIK_Euclid.ino - long click to exit overlay menu

Add longButton option to exit overlay (Save/Load/Preset)

modulove committed:
Update ARYTHMATIK_Euclid.ino - Add LongClick Mute in TOP_MENU

Now toggles mute status of the selected channel when in SETTING_TOP_MENU with long button push

awonak committed:
minor bug fixes; add missing encoder handler parameter, removed redundant disp_refresh set.

awonak committed:
avoid multiple calls to OLED_display by making force_refresh a global.

awonak committed:
fix setTextSize bug where preset remained 2, remove redundant setTextColor, minor readability fix.

modulove committed:
Last minute changes

Fix LGT8FX8P problem by limiting slots.
Also preset array with conditional size
Update rhythm patterns

modulove committed:
Nudge the sequencer when in TEMPO_MODE

Nudge the sequencer when in TEMPO_MODE

modulove committed:
Update drawTempo() 

display info when there's no external clock signal.
match style of the other overlay menus.

modulove committed:
Cosmetics ARYTHMATIK_Euclid.ino

New menu order. loved this 💯 
Few UI tweaks.

modulove committed:
Tweak performance. Display refresh only when necessary

Improve performance by reducing the frequency of display refreshes only when necessary.
Disables force_refresh if either of both clocks are running.

modulove committed:
Fix bootlogo & nudge without acceleration

Last chance for nudge feature fixed without acceleration?
Boologo fix (no more artifacts)

modulove committed:
some performance tweaks

removes some artifacts
removes random skips

modulove committed:
remove nudge feature

nope. no feature creep

modulove committed:
Change Encoder overlay menu handling

Long press loads / saves now
Remove UI elements to fight artifacts

modulove committed:
few minor optimizations

split OLED_display function.
early returns
smaller heart bootuplogo

modulove committed:
final stretch updates

encoder functions with switch statement
update rythms

modulove commented:
Ok here we go now. I have decided that the wait is over and spend enouch time with the firmware & testing. Cheers
